src/main/java/xades4j/providers/impl/HttpTimeStampTokenProvider.java

```diff
@@ -148,27 +55,41 @@ private InputStream getResponse(byte[] encodedRequest) throws TimeStampTokenGene
             out.write(encodedRequest);
             out.flush();
 
-            if (connection.getResponseCode() != HttpURLConnection.HTTP_OK)
-            {
+            if (connection.getResponseCode() != HttpURLConnection.HTTP_OK) {
                 throw new TimeStampTokenGenerationException(String.format("TSA returned HTTP %d %s", connection.getResponseCode(), connection.getResponseMessage()));
             }
 
-            // TODO do we need to invoke connection.disconnect()?
-            return new BufferedInputStream(connection.getInputStream());
-        } catch (IOException ex)
-        {
+            try (BufferedInputStream input = new BufferedInputStream(connection.getInputStream())) {
+                ByteArrayOutputStream baos = new ByteArrayOutputStream();
+                byte[] buffer = new byte[1024];
+                int len;
+                while ((len = input.read(buffer)) > -1) {
+                    baos.write(buffer, 0, len);
+                }
+                baos.flush();
+
+                return baos.toByteArray();
+            }
+        } catch (IOException ex) {
             throw new TimeStampTokenGenerationException("Error when connecting to the TSA", ex);
+        } finally {
+            if (connection != null) connection.disconnect();
         }
     }
 
-    private HttpURLConnection createHttpConnection() throws IOException
-    {
+    private HttpURLConnection createHttpConnection() throws IOException {
         HttpURLConnection connection = (HttpURLConnection) new URL(this.temporaryGetTSAUrl()).openConnection();
-        String username = this.tsaHttpData.getUsername(), password = this.tsaHttpData.getPassword();
+        String base64TsaUsrAndPwd;
+
+        if (this.tsaHttpData.getUsername() != null) {
```

Comment:
Why is this better than caching the final calculated value in an instance variable? Note that these providers can usually be singleton.